src/PureSMT/Process.hs

```diff
@@ -30,14 +29,18 @@ initZ3Instance ::
   Bool ->
   IO Solver
 initZ3Instance dbg = do
+  let ctxFinalizer =
+        [C.funPtr| void free_context(Z3_context ctx) {
+                                      Z3_del_context(ctx);
+                                      } |]
```

Comment:
The code here is a bit redundant. Is there a better way to do this ?

Reply:
I think I would have tried
```haskell
let ctxFinalizer ctx = [CU.block| void { Z3_del_context($(Z3_context ctx)); } |]
```

Reply:
> I think I would have tried
> 
> ```haskell
> let ctxFinalizer ctx = [CU.block| void { Z3_del_context($(Z3_context ctx)); } |]
> ```

It doesn't type:

```
Couldn't match type: GHC.Ptr.Ptr Z3.LogicalContext -> IO ()
                     with: GHC.Ptr.FunPtr (GHC.Ptr.Ptr Z3.LogicalContext -> IO ())
      Expected: FinalizerPtr Z3.LogicalContext
        Actual: GHC.Ptr.Ptr Z3.LogicalContext -> IO ()
```

Reply:
Another option that does type is
```haskell
ctxFinalizer <- [CU.exp| void (*fun_ptr)(Z3_context) { &Z3_del_context } |]
```
Any favorites ?

Reply:
I think I like the original more. You could have used this instead:
https://www.stackage.org/haddock/lts-19.30/base/Foreign-Concurrent.html#v:newForeignPtr

But I don't think it is better in any obvious way.

Reply:
Here there is a discussion that argues to keep the code as is: https://hackage.haskell.org/package/inline-c-0.9.1.6/docs/Language-C-Inline.html#v:funPtr